pkg/kubelet/kubelet.go

```diff
@@ -2737,6 +2747,139 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+// is created and updated for each new checkpoint. If a new checkpoint would
+// result in more than 'kl.maxContainerCheckpoints' the oldest checkpoint will
+// be removed.
+func (kl *Kubelet) handleMaxContainerCheckpoints(
+	podUID types.UID,
+	podFullName,
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+	checkpointTime int64,
+) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// set this to 10 as a default value if 0 or not set. The error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints needs to be > 0")
+	}
+
+	checkpointJsonName := fmt.Sprintf(
+		"checkpoint-%s-%s.json",
+		podFullName,
+		containerName,
+	)
+
+	checkpointJsonPath := filepath.Join(kl.getCheckpointsDir(), checkpointJsonName)
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.updateMaxContainerCheckpoints.Lock()
+	defer kl.updateMaxContainerCheckpoints.Unlock()
+
+	checkpointMetadata, err := func() (
+		*checkpointctl.KubernetesContainerCheckpointMetadata,
+		error,
+	) {
+		_, err := os.Stat(checkpointJsonPath)
+		// If the file does not exist return an empty structure.
+		if err != nil {
+			return &checkpointctl.KubernetesContainerCheckpointMetadata{}, nil
+		}
+		ccm := &checkpointctl.KubernetesContainerCheckpointMetadata{}
+		_, err = checkpointctl.ReadJSONFile(
+			ccm,
+			kl.getCheckpointsDir(),
+			checkpointJsonName,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to read checkpoint metadata file (%s): %w",
+				checkpointJsonPath,
+				err,
+			)
+		}
+		return ccm, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	fi, err := os.Stat(options.Location)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to stat checkpoint archive (%s): %w",
+			options.Location,
+			err,
+		)
+	}
+
+	numberOfCheckpoints := int32(len(checkpointMetadata.Checkpoints))
+
+	for numberOfCheckpoints >= kl.maxContainerCheckpoints {
+		klog.InfoS(
+			"More than maxContainerCheckpoints",
+			"container", containerName,
+			"numberOfCheckpoints", numberOfCheckpoints,
+			"maxContainerCheckpoints", kl.maxContainerCheckpoints,
+		)
+		// Find oldest checkpoint and delete it.
+		oldest := (checkpointMetadata.Checkpoints)[0].Timestamp
+		entryToDelete := 0
+		trimSizeBy := int64(0)
+		for i, checkpoint := range checkpointMetadata.Checkpoints {
+			if checkpoint.Timestamp < oldest {
+				oldest = checkpoint.Timestamp
+				entryToDelete = i
+				trimSizeBy = checkpoint.Size
+			}
+		}
+		klog.InfoS(
+			"Deleting container checkpoint archive",
+			"container", containerName,
+			"archive", (checkpointMetadata.Checkpoints)[entryToDelete].Archive,
+		)
+		// Ignoring error during delete in case the checkpoint archive
+		// has already been deleted.
+		_ = os.Remove((checkpointMetadata.Checkpoints)[entryToDelete].Archive)
```

Comment:
We should log the error with a higher verbosity from my point of view.

Reply:
also added.